### ISSUE TITLE
Appending new items to an empty list creates a wrong data-sid

### ISSUE THREAD

**Issue by StriderRanger** [dhilt/ngx-ui-scroll#107]
When appending new items to an empty list, I noticed some unwanted behavior:
- There is an empty space at the top of the list, represented by a `data-padding-backward` div with a height equals to the average height of an item in the list.
- the first item in the list has a `data-sid` equals to `2` instead of `1 `
- When scrolling backward the first item is duplicated

I think that the main problem is the wrong value that is attributed to the `data-sid` of the first element. I tried to see if there is something wrong with my datasource definition or the way that I am appending the items, but the source of the problem is not obvious.

I created [this stackblitz](https://stackblitz.com/edit/ngx-ui-scroll-append-index-prblm) where you can test the problem, and this is a screenshot of the results:
![append](https://user-images.githubusercontent.com/15088185/70655823-d386a700-1c50-11ea-80e1-60f31fb7155a.png)

Thanks!

**Comment by dhilt** [comment]
@StriderRanger Thanks again! This is definitely a bug. When UIScroll tries to append item(s) to empty buffer, he assumes that the `startIndex` had been used and assigns `startIndex + 1` as the first appending item index value... This particular method is responsible for the issue: 

https://github.com/dhilt/ngx-ui-scroll/blob/b1c90ac4914fbbbf00e91f0f94f55d852d9e03f4/src/component/classes/buffer.ts#L186-L188

Replacing it with 

```
  getIndexToAppend(eof?: boolean): number {
    return (!eof ? (this.items.length ? this.items[this.items.length - 1].$index : this.maxIndex - 1) : this.absMaxIndex) + 1;
  }
```

would fix the issue. But I need to consider if there are better ways, how to deal with prepend case and which tests should be written to cover this situation.

**Comment by StriderRanger** [comment]
Thanks! I will test the fix once it's available.

**Comment by dhilt** [comment]
@StriderRanger Hey! I just published v1.2.1 as "next" release, could you please try it ?

```
npm install ngx-ui-scroll@next
```

**Comment by StriderRanger** [comment]
Thank you for the update! I tested v1.2.1 on my stackblitz [example](https://stackblitz.com/edit/ngx-ui-scroll-append-index-prblm) and it worked fine, but in my local app I have always the same old behavior. I am not sure about the cause but perhaps it is due to the asynchronous way of processing the data.
I created [this example](https://stackblitz.com/edit/ngx-ui-scroll-ionic4-pagination-append-index) that reproduces the problem.

**Comment by dhilt** [comment]
@StriderRanger Yes... I missed eof/bof parameter case... In your last sample you call `append` with `eof = true`. I fixed it and released v1.2.2

Also, I would agree with you that datasource-append interaction is not clear in your sample. You requests 4 items on start via `Datasource`, returns empty result and then appends 7 items via `Adapter`. As I spoke before, I'd prefer to have all the logic inside the `Datasource.get` method: requesting N items via `Datasource` might asynchronously but straight enough result in N-items array, so no `append` is needed... like you did in [previous issue](https://github.com/dhilt/ngx-ui-scroll/issues/105).

Anyway, I believe this particular issue has been fixed after two phases, bit could you try v1.2.2 in your environment before?

**Comment by StriderRanger** [comment]
I tested the change in my environment and also in my stackblitz [example](https://stackblitz.com/edit/ngx-ui-scroll-ionic4-pagination-append-index-fix) and the `data-sid` was correct in both environments.

As you mentioned, In [#105](https://github.com/dhilt/ngx-ui-scroll/issues/105)'s [example](https://stackblitz.com/edit/ngx-ui-scroll-ionic4-pagination-workaround3) I changed all this pagination approach when seeing that the interaction with the data was not solid and that `Datasource.get` could handle all the process of paginating the data.

Thank you very much for your support, that was very helpful to better understand how to have a good implementation of the uiScroll  :)
